pkg/certificate/providers/mrc.go

```diff
@@ -0,0 +1,127 @@
+// `certificate.Provider`s from those MRCs
+type MRCComposer struct {
+	informerCollection *informers.InformerCollection
+	MRCProviderGenerator
+}
+
+// List returns the MRCs stored in the informerCollection's store
+func (m *MRCComposer) List() ([]*v1alpha2.MeshRootCertificate, error) {
+	if m.informerCollection == nil {
+		return nil, errNilInformerCollection
+	}
+	// informers return slice of pointers so we'll convert them to value types before returning
+	mrcPtrs := m.informerCollection.List(informers.InformerKeyMeshRootCertificate)
+	var mrcs []*v1alpha2.MeshRootCertificate
+	for _, mrcPtr := range mrcPtrs {
+		if mrcPtr == nil {
+			continue
+		}
+		mrc, ok := mrcPtr.(*v1alpha2.MeshRootCertificate)
+		if !ok {
+			continue
+		}
+		mrcs = append(mrcs, mrc)
+	}
+
+	return mrcs, nil
+}
+
+// Watch returns a channel that receives events whenever MRCs are added, updated, and deleted
+// from the informerCollection's MRC store. Channels returned from multiple invocations of
+// Watch() are unique and have no coordination with each other. Events are guaranteed
+// to be ordered for any particular resources, but NOT across different resources.
+// Watch() should only send new events as of the time the Watch call returns
+func (m *MRCComposer) Watch(ctx context.Context) (<-chan certificate.MRCEvent, error) {
+	if m.informerCollection == nil {
+		return nil, errNilInformerCollection
+	}
+
+	// To make sure we only send new events, we cache the object versions
+	// before the eventhandler is added
+	mrcPtrs := m.informerCollection.List(informers.InformerKeyMeshRootCertificate)
+	mrcObjectVersionCache := map[string]int64{}
+	for _, mrcPtr := range mrcPtrs {
+		if mrcPtr == nil {
+			continue
+		}
+		mrc, ok := mrcPtr.(*v1alpha2.MeshRootCertificate)
+		if !ok {
+			continue
+		}
+		mrcObjectVersionCache[mrcKeyFunc(*mrc)] = mrc.GetGeneration()
+	}
+
+	eventChan := make(chan certificate.MRCEvent)
+	m.informerCollection.AddEventHandler(informers.InformerKeyMeshRootCertificate, cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			mrc := obj.(*v1alpha2.MeshRootCertificate)
+			if !checkGeneration(mrcObjectVersionCache, *mrc) {
+				return
+			}
+			eventChan <- certificate.MRCEvent{
+				Type: certificate.MRCEventAdded,
+				MRC:  mrc,
+			}
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			oldMRC := oldObj.(*v1alpha2.MeshRootCertificate)
+			mrc := newObj.(*v1alpha2.MeshRootCertificate)
+			if !checkGeneration(mrcObjectVersionCache, *mrc) {
+				return
+			}
+			eventChan <- certificate.MRCEvent{
+				Type:   certificate.MRCEventUpdated,
+				MRC:    mrc,
+				OldMRC: oldMRC,
+			}
+		},
+		DeleteFunc: func(obj interface{}) {
+			mrc, ok := obj.(*v1alpha2.MeshRootCertificate)
+			// Sometimes, deleted objects will be of type DeletedFinalStateUnknown
+			// We don't send these events because we literally can't gain anything
+			// from knowing that an unknown MRC was deleted
```

Comment:
Does the object stay in the local cache in that scenario? If so, should we periodically prune the cache ourselves by querying the API so things like `List` remain accurate?

Reply:
To answer your question, yes the object would still be in the local cache during that scenario. Typically, the way to solve this is by setting a nonzero resync period on the informer OR the event handler. I ended up gutting this function so it's moot for this PR, but good question!